lib/src/util/environment_visitors.dart

```diff
@@ -0,0 +1,500 @@
+    _addLocalEnvironment();
+    _addElementToEnvironment(new _ConditionExpression(condition, value: false));
+    elseStatement?.accept(this);
+    return super._removeLocalEnvironment();
+  }
+
+  _visitIfStatement(IfStatement node) {
+    _addLocalEnvironment();
+    _addElementToEnvironment(new _ConditionExpression(node.condition));
+    node.condition?.accept(this);
+    node.thenStatement?.accept(this);
+    _removeLocalEnvironment();
+  }
+}
+
+class _ConditionExpression extends _ExpressionBox {
+  Expression expression;
+  bool value;
+
+  _ConditionExpression(this.expression, {this.value: true});
+
+  @override
+  bool haveToStop(Iterable<Element> elements) => false;
+
+  @override
+  String toString() => '$expression is $value';
+}
+
+class _DelegatingAstVisitor extends UnifyingAstVisitor {
+  final AstVisitor delegate;
+
+  _DelegatingAstVisitor(this.delegate);
+
+  @override
+  void visitNode(AstNode node) {
+    node.accept(delegate);
+    node.visitChildren(this);
+  }
+}
+
+class _ElementEnvironmentVisitor extends _EnvironmentVisitor<Element> {
+  _ElementEnvironmentVisitor(AstVisitor baseVisitor) : super(baseVisitor);
+
+  Element lookUp(String name) {
+    if (name == null) {
+      throw new ArgumentError.notNull('name');
+    }
+    for (final localEnvironment in environments) {
+      for (final element in localEnvironment) {
+        if (element.name == name) {
+          return element;
+        }
+      }
+    }
+    return null;
+  }
+
+  @override
+  visitClassDeclaration(ClassDeclaration node) {
+    _addLocalEnvironment();
+    node.members.forEach((e) {
+      _addElementToEnvironment(e.element);
+    });
+    _baseVisitor.visitClassDeclaration(node);
+    node.visitChildren(this);
+    _removeLocalEnvironment();
+  }
+
+  @override
+  visitCompilationUnit(CompilationUnit node) {
+    _addLocalEnvironment();
+    node.declarations.forEach((e) {
+      this._addElementToEnvironment(e.element);
+    });
+    _baseVisitor.visitCompilationUnit(node);
+    node.visitChildren(this);
+    _removeLocalEnvironment();
+  }
+
+  @override
+  visitFormalParameterList(FormalParameterList node) {
+    _baseVisitor.visitFormalParameterList(node);
+    node.parameterElements.forEach(this._addElementToEnvironment);
+    node.visitChildren(this);
+  }
+
+  @override
+  visitFunctionDeclaration(FunctionDeclaration node) {
+    // Necessary for local functions
+    if (node.parent is FunctionDeclarationStatement) {
+      _addElementToEnvironment(node.element);
+    }
+    _addLocalEnvironment();
+    _baseVisitor.visitFunctionDeclaration(node);
+    node.visitChildren(this);
+    _removeLocalEnvironment();
+  }
+
+  @override
+  visitVariableDeclaration(VariableDeclaration node) {
+    _baseVisitor.visitVariableDeclaration(node);
+    _addElementToEnvironment(node.element);
+    node.visitChildren(this);
+  }
+}
+
+abstract class _EnvironmentVisitor<E> extends _DelegatingAstVisitor {
+  final Queue<Queue<E>> environments = new Queue();
```

Comment:
> I saw ScopedVisitor and it is wonderful code ...

I actually think that it needs to be re-factored, but that's a different conversation. :-)

To use it you would create a subclass and implement `visitThisExpression` (similar to what you already have). When you get to the point of trying to resolve an identifier, you'd write something like
```
SimpleIdentifier identifier;
if (parent is PropertyAccess) {
  identifier = parent.propertyName;
} else if (parent is MethodInvocation) {
  identifier = parent.methodName;
} else {
  return null;
}
Element element = nameScope.lookup(identifier, referencingLibrary);
```
where the `referencingLibrary` is the `LibraryElement` associated with the compilation unit being traversed.